libc/unistd_file.c

```diff
@@ -0,0 +1,848 @@
+static void assert_read_more(const int fildes, const char *dest, const int n)
+{
+	int temp;
+	int cnt = 0;
+	int nread = 0;
+
+	/* Assuming that n attempts would be sufficient */
+	while (nread < n && cnt++ < n) {
+		/* Always try to read 2 times more than it's possible */
+		temp = read(fildes, buf, 2 * n);
+		TEST_ASSERT_GREATER_OR_EQUAL(0, temp);
+		nread += temp;
+	}
+	TEST_ASSERT_EQUAL(n, nread);
+}
+
+
+TEST_GROUP(unistd_file);
+
+
+TEST_SETUP(unistd_file)
+{
+	/* open a file */
+	fd = open(FNAME, O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fd);
+	memset(buf, '\0', sizeof(buf));
+}
+
+
+TEST_TEAR_DOWN(unistd_file)
+{
+	memset(buf, '\0', sizeof(buf));
+	if (fd >= 0) {
+		TEST_ASSERT_EQUAL_INT(0, close(fd));
+	}
+	TEST_ASSERT_EQUAL_INT(0, remove(FNAME));
+}
+
+
+TEST(unistd_file, file_close)
+{
+	/* close an opened descriptor */
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+
+	/* try to close a closed descriptor */
+	TEST_ASSERT_EQUAL_INT(-1, close(fd));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	/* this test closes fd by itself, invalidate it */
+	fd = -1;
+}
+
+
+/* Simple usage of read() and write() */
+TEST(unistd_file, file_readwrite_nbytes)
+{
+	assert_write(fd, LINE1);
+
+	/* repoen file readonly */
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+	fd = open(FNAME, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fd);
+	{
+		/* try to read LINE1 from file */
+		assert_read(fd, buf, sizeof(LINE1) - 1);
+		TEST_ASSERT_EQUAL_STRING(LINE1, buf);
+
+		/* assert EOF */
+		TEST_ASSERT_EQUAL_INT(0, read(fd, buf, sizeof(LINE1) - 1));
+	}
+}
+
+
+TEST(unistd_file, file_write_zero)
+{
+	/* test writing no bytes */
+	TEST_ASSERT_EQUAL_INT(0, write(fd, NULL, 0));
+}
+
+
+/* simultaneous write with two open descriptors */
+TEST(unistd_file, file_write_reopened)
+{
+	int sum;
+	int fdr;
+
+	/* reopen file descriptor and open another for reading */
+	fd2 = open(FNAME, O_WRONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fd2);
+	fdr = open(FNAME, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fdr);
+	{
+		/* write to the same file consecutively from two different file descriptors */
+		assert_write(fd, LINE1);
+		assert_write(fd2, LINE2);
+
+		/* try to read too much for simple error detection */
+		assert_read_more(fdr, buf, sizeof(LINE2) - 1);
+		TEST_ASSERT_EQUAL_STRING(LINE2, buf);
+
+		/* test for EOF */
+		TEST_ASSERT_EQUAL_INT(0, read(fdr, buf, sum));
+	}
+	TEST_ASSERT_EQUAL_INT(0, close(fdr));
+	TEST_ASSERT_EQUAL_INT(0, close(fd2));
+}
+
+
+/* test simultaneous write() on duplicated file descriptor */
+TEST(unistd_file, file_write_dup)
+{
+	int sum;
+	int fdr;
+
+	/* duplicate file descriptor and open another for reading */
+	fd2 = dup(fd);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd2);
+	fdr = open(FNAME, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fdr);
+	{
+		/* write to the same file consecutively from descriptor and its copy */
+		assert_write(fd, LINE1);
+		assert_write(fd2, LINE2);
```

Comment:
I don't really like that `assert_write` remove NULL terminator from strings and null terminator is added by `memset`ing `buff` - this makes code using it somewhat less intuitive and error-prone

Reply:
Sorry, stayed by accident, not true